ENH Support for sparse matrices added to `sklearn.metrics.silhouette_samples`

awinml opened the pull request:
#### Reference Issues/PRs
Fixes #18524 
Fixes #18723 


#### What does this implement/fix? Explain your changes.
Support for sparse matrices added to `sklearn.metrics.silhouette_samples`
Suggestions from the original stalled PR were implemented.
> The changes update the reduce function used for computing the intra-cluster and inter-cluster distances. The current version is failing at, a) the pre-computed check for sparse matrices while getting the diagonal elements b) when trying to index a sparse matrix to pass weights to np.bincount function

- Multiple merge conflicts with main were resolved.
- New test `test_silhouette_samples_sparse` was added to check for sparse matrices and the implementation of `silhouette_samples` with sparse matrices.
- Test `test_silhouette_sparse_input` has been removed.
- Check for non-zero diagonal entries in `silhouette_samples` was updated to work with sparse matrices.

Commit:
Updated silhouette reduce function

Commit:
Fixed lint tests

Commit:
Codecov coverage failing test

Commit:
Fixed indentation to conform to PEP style

Commit:
Fixed coding style

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into feature/18524
Merge changes from remote

Commit:
Unit test for changes

Commit:
Avoid converting sprase matrix to dense matrix

Commit:
Fixed line spaces in test file:

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into feature/18524

Commit:
Unit test for sparse input implementation

awinml committed:
Resolved Merge Conflicts in metrics/cluster/_unsupervised.py

awinml committed:
Updated  with suggested change

awinml committed:
Updated check for non-zero diagonal entries in silhouette_samples

awinml committed:
Fixed naming for error message

awinml committed:
Update changelog 1.2

thomasjpfan commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -282,6 +285,27 @@ def test_silhouette_nonzero_diag(dtype):
 
 
+@pytest.mark.parametrize("to_sparse", (csr_matrix, csc_matrix))
+def test_silhouette_sparse_input(to_sparse):
```
Currently, this test is only smoke test, because it does not assert anything about the output.

I think we can parameterize `test_silhouette_sparse_implementation` to check both `csr_matrix` and `csc_matrix` works and remove `test_silhouette_sparse_input`.

thomasjpfan commented on sklearn/metrics/cluster/_unsupervised.py:
```diff
@@ -131,15 +132,24 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
     # accumulate distances from each sample to each cluster
-    clust_dists = np.zeros((len(D_chunk), len(label_freqs)), dtype=D_chunk.dtype)
-    for i in range(len(D_chunk)):
+    clust_dists = np.zeros((n_chunk_samples, len(label_freqs)), dtype=D_chunk.dtype)
+    for i in range(n_chunk_samples):
+        if issparse(D_chunk):
```
So we do not need to call `issparse` many times:

```suggestion
    clust_dists = np.zeros((n_chunk_samples, len(label_freqs)), dtype=D_chunk.dtype)
    D_chunck_is_sparse = issparse(D_chunk)
    for i in range(n_chunk_samples):
        if D_chunck_is_sparse:
```

thomasjpfan reviewed (commented):
Thank you for the PR!

awinml committed:
Removed test_silhouette_sparse_input, paramterized test_silhouette_samples_sparse

awinml commented:
The following changes have been made @thomasjpfan :
- Parameterized `test_silhouette_samples_sparse` to check for both `csr_matrix` and `csc_matrix`. 
- Removed `test_silhouette_sparse_input`.
- Modified check so that `issparse` is called only once.

awinml committed:
Removed Unnecessary imports

awinml commented:
Could you please kindly review the changes? @lorentzenchr @jeremiedbb @thomasjpfan

awinml committed:
Merge branch 'main' into silhouette_samples_sparse_matrices

awinml commented:
Would you please kindly review this PR @glemaitre ?

awinml committed:
Merge branch 'main' into silhouette_samples_sparse_matrices

awinml committed:
Merge branch 'main' into silhouette_samples_sparse_matrices

awinml commented:
Re-requesting review @thomasjpfan @glemaitre.

thomasjpfan commented on doc/whats_new/v1.2.rst:
```diff
@@ -533,6 +533,11 @@ Changelog
   :func:`metrics.pairwise.manhattan_distances` is deprecated and will be removed in 1.4.
   :pr:`24630` by :user:`Rushil Desai <rusdes>`.
+  
+- |Enhancement| :class:`metrics.silhouette_samples` nows accepts a sparse
```
Since 1.2 is released, this change log entry needs to be moved to 1.3

thomasjpfan commented on sklearn/metrics/cluster/_unsupervised.py:
```diff
@@ -131,15 +132,25 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
+    D_chunk_is_sparse = issparse(D_chunk)
+    for i in range(n_chunk_samples):
+        if D_chunk_is_sparse:
+            current_chunk = D_chunk.getrow(i)
```
It is unfortunate that `getrow` always returns a copy, even for CSR matrices: https://docs.scipy.org/doc//scipy/reference/generated/scipy.sparse.csr_array.getrow.html?highlight=getrow#scipy.sparse.csr_array.getrow

For me, I am okay with optimizing for `csr`, because most of scikit-learn is optimized for CSR matrices:

```python
    if D_chunk_is_sparse and D_chunk.format != "csr":
        D_chunk = D_chunk.tocsr()

    for i in range(n_chunk_samples):
        if D_chunk_is_sparse:
            indptr = D_chunk.indptr
            indices = D_chunk.indices[indptr[i] : indptr[i + 1]]
            sample_weights = D_chunk.data[indptr[i] : indptr[i + 1]]
            sample_labels = np.take(labels, indices)
```

If the precomputed matrix is CSR, `tocsr` is a no-op and getting the `sample_labels` and `sample_weights` do not make a copy anymore.

awinml replied:
Okay. I will make the necessary changes, so that we optimize for CSR matrices.

awinml committed:
Merge branch 'main' into silhouette_samples_sparse_matrices

awinml committed:
Update changelog

awinml committed:
Optimize for CSR

awinml committed:
Merge branch 'main' into silhouette_samples_sparse_matrices

awinml committed:
Update sklearn/metrics/cluster/_unsupervised.py

awinml committed:
Add test for Non-CSR sparse matrices

awinml commented on sklearn/metrics/cluster/_unsupervised.py:
```diff
@@ -131,15 +132,30 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
+    for i in range(n_chunk_samples):
+        if D_chunk_is_sparse:
+            if D_chunk.format != "csr":
+                D_chunk = D_chunk.tocsr()
```
@thomasjpfan I already have a test that takes in sparse non-csr inputs. Codecov still flags this line as not covered by tests. How should I go about increasing coverage?

thomasjpfan replied:
`silhouette_samples` already converts all data to `csr` through `pairwise_distances_chunked` before passing it to `_silhouette_reduce`. This means we can remove this `D_chunk.tocsr()`.

I'll also update the docstring for `_silhouette_reduce` to state that `D_chunk` can be a spares csr matrix.

awinml replied:
I will make the required changes. This should also solve the coverage issues.

awinml committed:
Merge branch 'main' into silhouette_samples_sparse_matrices

thomasjpfan commented on doc/whats_new/v1.3.rst:
```diff
@@ -147,6 +147,17 @@ Changelog
+:mod:`sklearn.metrics`
+......................
+
+- |Enhancement| :class:`metrics.silhouette_samples` nows accepts a sparse
```
Can this entry be moved down to the `sklearn.metrics` section below?

awinml replied:
Yes. I must have missed that when resolving the merge conflicts.

thomasjpfan commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -282,6 +284,32 @@ def test_silhouette_nonzero_diag(dtype):
+    assert_allclose(sparse_out, dense_out)
+
+
+@pytest.mark.parametrize("to_sparse", (csc_matrix, dok_matrix, lil_matrix))
```
We can remove the above test by adding `csr_matrix` here.

awinml replied:
Yeah, I will do that. This test becomes redundant after removing the check for `D_chunk.tocsr()`.

thomasjpfan reviewed (commented):
Thanks for the update!

awinml committed:
Update Changelog

awinml committed:
Remove redundant csr conversion

thomasjpfan reviewed (approved):
Can you update the following docstrings to state that sparse data is supported?

```patch
diff --git a/sklearn/metrics/cluster/_unsupervised.py b/sklearn/metrics/cluster/_unsupervised.py
index ce7dff9747..14d364e4db 100644
--- a/sklearn/metrics/cluster/_unsupervised.py
+++ b/sklearn/metrics/cluster/_unsupervised.py
@@ -123,7 +123,7 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
 
     Parameters
     ----------
-    D_chunk : array-like of shape (n_chunk_samples, n_samples)
+    D_chunk : {array-like, sparse CSR matrix} of shape (n_chunk_samples, n_samples)
         Precomputed distances for a chunk.
     start : int
         First index in the chunk.
@@ -190,7 +190,7 @@ def silhouette_samples(X, labels, *, metric="euclidean", **kwds):
 
     Parameters
     ----------
-    X : array-like of shape (n_samples_a, n_samples_a) if metric == \
+    X : {array-like, sparse matrix} of shape (n_samples_a, n_samples_a) if metric == \
             "precomputed" or (n_samples_a, n_features) otherwise
         An array of pairwise distances between samples, or a feature array.
```

Otherwise, LGTM

awinml committed:
Update sklearn/metrics/cluster/tests/test_unsupervised.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

awinml committed:
Update docstrings for sparse data support

awinml commented:
I have updated the docstrings. Thanks!

glemaitre commented on sklearn/metrics/cluster/_unsupervised.py:
```diff
@@ -174,7 +187,7 @@ def silhouette_samples(X, labels, *, metric="euclidean", **kwds):
     Parameters
     ----------
-    X : array-like of shape (n_samples_a, n_samples_a) if metric == \
+    X : {array-like, sparse matrix} of shape (n_samples_a, n_samples_a) if metric == \
```
There is something quite difficult to tract in the code here. We accept both CSC and CSR matrix but the reduce function only works on CSR matrix. Looking at the code, the `pairwise_distance` will make the conversion from CSC to CSR but I think it would be better to only accept CSR implying that we convert `X` to CSR when other formats at the first `check_array`.

It is more explicit.

awinml replied:
Yes, that's a good point. I will update the code to only accept CSR sparse matrices.

glemaitre commented on sklearn/metrics/cluster/_unsupervised.py:
```diff
@@ -131,15 +132,27 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
+    D_chunk_is_sparse = issparse(D_chunk)
+
+    for i in range(n_chunk_samples):
+        if D_chunk_is_sparse:
```
I don't really like the fact that we do a check for sparsity within the `for` loop.
I would inverse both:

```python
if issparse(D_chunk):
    for ...
else:
    for ...
```

glemaitre replied:
also I would add a check that the matrix is for CSR format:

```python
if issparse(D_chunck):
    if D_chunk.format != "csr":
        raise TypeError(...)
    for ...
...
```

glemaitre commented on sklearn/metrics/cluster/_unsupervised.py:
```diff
@@ -131,15 +132,27 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
     # accumulate distances from each sample to each cluster
-    clust_dists = np.zeros((len(D_chunk), len(label_freqs)), dtype=D_chunk.dtype)
-    for i in range(len(D_chunk)):
+    clust_dists = np.zeros((n_chunk_samples, len(label_freqs)), dtype=D_chunk.dtype)
```
Since we modify the code, let's be explicit and and call this `cluster_distances`

awinml replied:
Okay, I will do that.

glemaitre commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -19,8 +21,8 @@ def test_silhouette():
-    X_dok = sp.dok_matrix(X_dense)
-    X_lil = sp.lil_matrix(X_dense)
+    X_dok = dok_matrix(X_dense)
+    X_lil = lil_matrix(X_dense)
```
We should also have a CSC matrix here.

glemaitre commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -282,6 +284,19 @@ def test_silhouette_nonzero_diag(dtype):
 
+@pytest.mark.parametrize("to_sparse", (csr_matrix, csc_matrix, dok_matrix, lil_matrix))
+def test_silhouette_samples_sparse(to_sparse):
+    """Check that silhouette_samples works for sparse matrices correctly"""
```
I think that we should specified that this is with the precomputed metric.
I think that we should have an additional test without precomputed metric.

awinml replied:
I have added the additional test with the default Euclidean metric.

awinml committed:
Rename clust_dist to cluster_distances

awinml committed:
Removed sparse check from loop

awinml committed:
Accept only CSR matrices

awinml committed:
Add new test for eucidean metric

awinml committed:
Merge branch 'main' into silhouette_samples_sparse_matrices

awinml commented:
@glemaitre All the comments have been addressed with the respective changes. Would you kindly please have a look at the updated code?

glemaitre commented on sklearn/metrics/cluster/_unsupervised.py:
```diff
@@ -122,7 +123,7 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
     Parameters
     ----------
-    D_chunk : array-like of shape (n_chunk_samples, n_samples)
+    D_chunk : {array-like, sparse CSR matrix} of shape (n_chunk_samples, n_samples)
```
```suggestion
    D_chunk : {array-like, sparse matrix} of shape (n_chunk_samples, n_samples)
```

glemaitre replied:
This change is to agree with our docstyle guideline.

glemaitre commented on sklearn/metrics/cluster/_unsupervised.py:
```diff
@@ -122,7 +123,7 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
     ----------
-    D_chunk : array-like of shape (n_chunk_samples, n_samples)
+    D_chunk : {array-like, sparse CSR matrix} of shape (n_chunk_samples, n_samples)
         Precomputed distances for a chunk.
```
```suggestion
        Precomputed distances for a chunk. When providing a sparse matrix, a
        CSR format should be provided to avoid an additional copy.
```

glemaitre commented on sklearn/metrics/cluster/_unsupervised.py:
```diff
@@ -174,7 +196,7 @@ def silhouette_samples(X, labels, *, metric="euclidean", **kwds):
-    X : array-like of shape (n_samples_a, n_samples_a) if metric == \
+    X : {array-like, sparse matrix} of shape (n_samples_a, n_samples_a) if metric == \
             "precomputed" or (n_samples_a, n_features) otherwise
         An array of pairwise distances between samples, or a feature array.
```
```suggestion
        An array of pairwise distances between samples, or a feature array. If
        a sparse matrix is provided, CSR format should be favoured avoiding
        an additional copy.
```

awinml committed:
Fix docstrings

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

awinml committed:
Merge branch 'main' into silhouette_samples_sparse_matrices

awinml committed:
Add test for _silhouette_reduce

glemaitre commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -282,6 +286,49 @@ def test_silhouette_nonzero_diag(dtype):
+
+@pytest.mark.parametrize("to_non_csr_sparse", (csc_matrix, dok_matrix, lil_matrix))
+def test_silhouette_reduce(to_non_csr_sparse):
+    """
+    Check for non-CSR input to private method `_silhouette_reduce`.
+    """
```
```suggestion
    """Check for non-CSR input to private method `_silhouette_reduce`."""
```

glemaitre committed:
Update sklearn/metrics/cluster/tests/test_unsupervised.py

glemaitre reviewed (approved):
I checked locally if the `_unsupervised.py` file was all covered which is the case. So LGTM.

glemaitre commented:
Merging. Thanks @awinml
